STL_Extension/include/CGAL/Handle_for.h

```diff
@@ -149,12 +149,14 @@ class Handle_for
         return *this;
     }
 
-    ~Handle_for()
+    ~Handle_for() noexcept
     {
-      if (--(ptr_->count) == 0) {
-        Allocator_traits::destroy(allocator, ptr_);
-        allocator.deallocate( ptr_, 1);
-      }
+      try{
+        if (--(ptr_->count) == 0) {
+          Allocator_traits::destroy(allocator, ptr_);
+          allocator.deallocate( ptr_, 1);
+        }
+      } catch(...) {}
```

Comment:
I am not going to fight it, but I want to mention that I am not fond of this new code
- it is ugly
- it goes against the default specified by the language

Reply:
Many of the destructors in the CGAL codebase can throw because of a CGAL_assertion, so when it calls deallocate here the program will terminate. (presumably, that's not desirable, that's what the change intends to prevent)

 In most places, this is handled by a `CGAL_destructor_assertion_catch` which preserves the exception when possible. But here because the noexcept is enforced by a base class we can only swallow the exception.

However you might agree that the correct way to solve all these problems, is never to throw from a destructor in the first place. Then there wouldn't have to be workarounds for the default behaviour specified by the language.

Reply:
> Many of the destructors in the CGAL codebase can throw because of a CGAL_assertion, so when it calls destroy/deallocate here the program will terminate. (presumably, that's not desirable, that's what the change intends to prevent)

It isn't obvious that it isn't desirable. The standard didn't gratuitously introduce `std::terminate`, it was better in those circumstances than data corruption or security holes.

> In most places, this is handled by a `CGAL_destructor_assertion_catch` which preserves the exception when possible. But here because the noexcept is enforced by a base class we can only swallow the exception.
> 
> However you might agree that the correct way to solve all these problems, is never to throw from a destructor in the first place. Then there wouldn't have to be workarounds for the default behaviour specified by the language.

And assertions are not supposed to trigger, so that's good, no need to introduce workarounds for the case where they trigger :wink:
Presumably, we aren't checking preconditions on user input in those destructors. If we assert, it probably means a bug in CGAL, or a side effect of some memory corruption, anyway something bad. We can disable those checks, as your patch does, that makes your checker happy and possibly fails to detect an issue. Or we can have checks that abort if they cannot throw exceptions, which is what the language specifies.

I wonder if separating more clearly debug assertions from checks that can make sense for users would reduce the number of destructors affected.

Some assertions are meant to protect from users who don't know what they are doing, say a user constructing Interval_nt(3,2). CGAL never constructs an interval with a>b, but since CGAL and the user's code call the same constructor, they go through the same assertion, and it looks like the CGAL could throw although it won't. If this happens in a noexcept function (like a destructor), the compiler has to generate instructions on what to do with that non-existing exception, for instance ignore it or abort. The amount of code is essentially the same, there shouldn't be much impact on performance, and since the exception never happens, the behavior is also the same. One code is simpler (no need for try-catch), and on the off chance some memory corruption causes the assertion to trigger, it will safely abort instead of executing random instructions.

Anyway, I said I won't fight...

Reply:
I appreciate your input @mglisse . It's always worthwhile having a discussion even if there are differences in opinion.

Reply:
> Presumably, we aren't checking preconditions on user input in those destructors. 

Not a destructor, but here is a noexcept method with a `CGAL_precondition`

https://github.com/CGAL/cgal/blob/master/STL_Extension/include/CGAL/Handle.h#L70-L79

Reply:
> Not a destructor, but here is a noexcept method with a `CGAL_precondition`

Yes, I think I remember adding this noexcept and wondering about the precondition. In the end, I was ok with having the program terminate if a null pointer was found in debug mode. The precondition isn't there in release mode. Not testing in debug mode would be a regression. Making the function conditionally noexcept would have been an option, but it can change the path taken by the code, and it felt more useful to lie a bit and have debug mode test something closer to release mode, than to insist on propagating the error as an exception. Of course different opinions are always possible.

Reply:
>Making the function conditionally noexcept 

In our usage of the CGAL library (RapCAD and OpenSCAD) we enable CGAL_precondition in release mode, so this would be more suitable for us I think.

Reply:
Can't you check the prerequisites in inputs before calling the functions? Or is it too expensive compared to do a try-catch?

Reply:
> Can't you check the prerequisites in inputs before calling the functions?
Not sure I follow what you mean @sloriot